### PR TITLE
[Core] Eliminate parallel worker per-step task scheduling overhead

### PR DESCRIPTION
**Pull request by njhill**
This PR replaces #3763.

Common logic is handled in the `DistributedGPUExecutor` superclass and used by both Ray and Mutliprocessing executors.

### CODE REVIEW

**Comment by njhill** [comment]
This is failing the distributed spec decoding test - TP spec decoding was added since the original PR and I need to look more closely at the flow for it and how to best integrate these changes with that.

**Comment by rkooo567** [comment]
we found nccl broadcasting overhead is very big now for high tp, and this PR is very important for reducing the gap. I will take a look at the PR tmrw!

**Comment by njhill** [comment]
Thanks @rkooo567! This PR actually doesn't reduce any nccl broadcast overhead, it just eliminates most of the non-torch.distributed RPCs which turn out to be much more significant (see the measurements in the [original PR](https://github.com/vllm-project/vllm/pull/3763#issue-2217188554)).

We can/should additionally reduce the amount of nccl broadcasting done, #4844 is a first step, I'm working on the next steps now.

I'll try to address the above-mentioned spec decoding TP issue with this PR today.

File: vllm/engine/async_llm_engine.py [hunk 1]
```diff
@@ -234,6 +234,10 @@ async def step_async(
         # Log stats.
         self.do_log_stats(scheduler_outputs, output)
 
+        if not request_outputs:
```
Comment (rkooo567) [inline vllm/engine/async_llm_engine.py:237]: QQ: why is this necessary?

File: vllm/worker/worker.py [hunk 2]
```diff
@@ -282,6 +270,26 @@ def execute_model(
         # to conform to interface.
         return [output]
 
+    @torch.inference_mode()
+    def start_worker_execution_loop(self) -> None:
+        """Execute model loop in parallel worker."""
```
Comment (rkooo567) [inline vllm/worker/worker.py:275]:
```suggestion
        """Execute model loop in parallel worker. You can stop the loop by executing a driver worker with an empty output. See `stop_remote_worker_execution_loop` for more details """
```

File: vllm/executor/ray_gpu_executor.py [hunk 3]
```diff
@@ -171,23 +173,17 @@ def _init_workers_ray(self, placement_group: "PlacementGroup",
                           max_concurrent_workers=self.parallel_config.
                           max_parallel_loading_workers)
 
-    def execute_model(
+    def _driver_execute_model(
```
Comment (rkooo567) [inline vllm/executor/ray_gpu_executor.py:176]: Can you add a docstring executing this with an empty input will stop worker execution loop?

File: vllm/executor/distributed_gpu_executor.py [hunk 4]
```diff
@@ -104,23 +125,35 @@ def _run_workers(
...
+                self._start_worker_execution_loop())
+
+        # Only the driver worker returns the sampling results.
+        return await self._driver_execute_model_async(execute_model_req)
+
+    async def stop_remote_worker_execution_loop_async(self) -> None:
+        if self.parallel_worker_tasks is None:
+            return
+
+        await self._driver_execute_model_async()
```
Comment (rkooo567) [inline vllm/executor/distributed_gpu_executor.py:143]: Add comment this API empty input will stop the execution loop from all workers.

**Review by rkooo567 — APPROVED** [review]
Hmm wonder if there's a cleaner way to stop the execution loop... (seems like too much impl details are leaked) unfortunately, I couldn't come up with a very good idea, so right now it is probably the best to document things more verbosely.

**Comment by rkooo567** [comment]
Btw, I also talked with @cadedaniel. I think this optimization makes spec decoding pretty complicated (and I feel like PP as well). @njhill I am curious if there's a way to disable this when spec decoding is enabled for the short term? I think we will need to discuss more holistic arcthitecture that will work for all different setup (spec decoding/pp/tp) cleanly, but this seems important for the short term at least

**Comment by andoorve** [comment]
> Btw, I also talked with @cadedaniel. I think this optimization makes spec decoding pretty complicated (and I feel like PP as well). @njhill I am curious if there's a way to disable this when spec decoding is enabled for the short term in a clean way?

+1 for PP @rkooo567

File: vllm/executor/distributed_gpu_executor.py [hunk 5]
```diff
@@ -1,18 +1,29 @@
...
 logger = init_logger(__name__)
 
 
 class DistributedGPUExecutor(GPUExecutor):
     """Abstract superclass of multi-GPU executor implementations."""
 
+    def __init__(self, *args, **kwargs):
+        # This is non-None when the execute model loop is running
+        # in the parallel workers
+        self.parallel_worker_tasks: Optional[Union[Any, Awaitable[Any]]] = None
```
[the remark this answers is not in the document]
Comment (njhill) [reply vllm/executor/distributed_gpu_executor.py:20]: I think this was to keep mypy happy. It's awaitable in the `AsyncLLMEngine` case but not in `LLMEngine` case. I can add a comment for that.

File: vllm/executor/distributed_gpu_executor.py [hunk 6]
```diff
@@ -88,13 +103,19 @@ def save_sharded_state(
...
+        raise NotImplementedError
+
     @abstractmethod
     def _run_workers(
         self,
         method: str,
         *args,
-        driver_args: Optional[Tuple[Any, ...]] = None,
-        driver_kwargs: Optional[Dict[str, Any]] = None,
+        remote_workers_only_async: bool = False,
```
[the remark this answers is not in the document]
Comment (njhill) [reply vllm/executor/distributed_gpu_executor.py:118]: Yes that's exactly right, and I was finding it hard to think of a good not-excessively-long name. `async_run_remote_workers_only` sounds better, I'll rename and add a clear explanation.

File: vllm/executor/multiproc_gpu_executor.py [hunk 7]
```diff
@@ -92,15 +98,13 @@ def _run_workers(
...
 
-        if driver_args is None:
-            driver_args = args
-        if driver_kwargs is None:
-            driver_kwargs = kwargs
+        if remote_workers_only_async:
+            # Just return futures
+            return worker_outputs
 
         # Start the driver worker after all the ray workers.
```
[the remark this answers is not in the document]
Comment (njhill) [reply vllm/executor/multiproc_gpu_executor.py:105]: It was there before, but I agree, will remove!

**Comment by njhill** [comment]
Thanks for the great review comments @rkooo567!

> Btw, I also talked with @cadedaniel. I think this optimization makes spec decoding pretty complicated (and I feel like PP as well). @njhill I am curious if there's a way to disable this when spec decoding is enabled for the short term in a clean way?

@rkooo567 @andoorve this would definitely be an option but would also come at the cost of at least some additional complexity. I haven't yet dug in enough but I'm hopeful about adapting it to work with spec decoding too. That could be misplaced optimism and I'm less sure about PP. I'll first look at those and if the complexity is too great then changing it to work only in non- spec decoding / PP case could be the backup plan.

I was actually hoping this could be a stepping-stone to elimination of the "control plane" RPCs altogether. The workers could remain in a permanent torch.distributed broadcast loop, with keepalives when idle if timeouts are an issue.

File: vllm/engine/async_llm_engine.py (hunk 1 above)
Comment (njhill) [reply vllm/engine/async_llm_engine.py:237]: @rkooo567 do you mean why only stop if `request_outputs` is empty? Or why do we need to stop the worker loop at all?

Comment (rkooo567) [reply vllm/engine/async_llm_engine.py:237]: oh the latter! """why do we need to stop the worker loop at all?"""

**Comment by andoorve** [comment]
> I was actually hoping this could be a stepping-stone to elimination of the "control plane" RPCs altogether. The workers could remain in a permanent torch.distributed broadcast loop, with keepalives when idle if timeouts are an issue.

@njhill This is a great point, I also committed a change to PyTorch in anticipation of these kinds of optimizations (Replace control-place RPC with torch distributed) for PP in the future: https://github.com/pytorch/pytorch/commit/b96b1e8cff029bb0a73283e6e7f6cc240313f1dc

It's a real pain to get multiple sends/recvs on a single rank though, so this is definitely a more future item once we get the basic functionality of PP working.

**Comment by rkooo567** [comment]
@njhill btw, this was the major bottleneck for us now, so lmk if there's any way I can help accelerating the PR merge!!

**Comment by njhill** [comment]
@rkooo567 about to update now, will push within next couple of hours hopefully!

File: vllm/engine/async_llm_engine.py (hunk 1 above)
Comment (njhill) [reply vllm/engine/async_llm_engine.py:237]: Well ultimately I'm hoping we won't need to stop it as mentioned [below](https://github.com/vllm-project/vllm/pull/4894#issuecomment-2121679761).

But I think that would be a bigger leap from how things work now. In particular the mechanism for other worker control RPCs would also need to be re-worked (like add/remove lora adapters), and we would need to figure out timeout issues since I don't think the collective ops are intended for long waits.

**Comment by rkooo567** [comment]
@njhill let me know when it is ready to review again!

**Comment by njhill** [comment]
Thanks @rkooo567, I made some updates but had to deal with other things for the remainder of the day. If I don't get a chance to finish debugging tonight I'll do it first thing in the morning (pacific time)

**Review by cadedaniel — APPROVED** [review]

**Comment by cadedaniel** [comment]
Approving spec decode changes

File: vllm/engine/async_llm_engine.py (hunk 1 above)
Comment (rkooo567) [reply vllm/engine/async_llm_engine.py:237]: It'd be great if you can comment what edge cases are handled from the comment here by doing this!

File: vllm/executor/distributed_gpu_executor.py [hunk 8]
```diff
@@ -88,39 +114,84 @@ def save_sharded_state(
...
+        Args:
+            async_run_remote_workers_only: If True the method will be run only
+                in the remote workers, not the driver worker. It will also be
+                run asynchronously and return a list of futures rather than
+                blocking on the results.
+        """
+        raise NotImplementedError
+
+    @abstractmethod
+    def _wait_for_tasks_completion(self, parallel_worker_tasks: Any) -> None:
```
Comment (rkooo567) [inline vllm/executor/distributed_gpu_executor.py:149]: QQ: this can hang forever if the loop is not finished properly. Should we allow timeout here and kill workers if timeout is reached with an exception? (theoretically, I think if it takes > 30s, I think there's something wrong).

File: vllm/spec_decode/spec_decode_worker.py [hunk 9]
```diff
@@ -231,75 +231,61 @@ def initialize_cache(self, num_gpu_blocks: int,
...
-                    execute_model_req.seq_group_metadata_list) == 0:
-                return self._run_no_spec(execute_model_req,
-                                         skip_proposer=disable_all_speculation)
-
-            return self._run_speculative_decoding_step(execute_model_req,
-                                                       num_lookahead_slots)
-        else:
-            self._run_non_driver_rank(num_lookahead_slots)
+        if execute_model_req is None:
+            # No data to run, notify other workers to stop the execution loop.
```
Comment (rkooo567) [inline vllm/spec_decode/spec_decode_worker.py:246]: can you add more details here? I think it is a little difficult to understand without enough context.

```
all workers are running infinite loop with broadcast_tensor_dict, and it stops the loop when a driver broadcasts an empty input. Send an empty input to notify all other workers to stop execution loop. 
```

**Review by rkooo567 — APPROVED** [review]
LGTM if tests pass!

**Comment by andoorve** [comment]
Hey @njhill, did you get a chance to look at this? I feel like this would cause quite a few changes to PP in its current form.

> @rkooo567 @andoorve this would definitely be an option but would also come at the cost of at least some additional complexity. I haven't yet dug in enough but I'm hopeful about adapting it to work with spec decoding too. That could be misplaced optimism and I'm less sure about PP. I'll first look at those and if the complexity is too great then changing it to work only in non- spec decoding / PP case could be the backup plan.

File: vllm/executor/distributed_gpu_executor.py (hunk 8 above)
Comment (njhill) [reply vllm/executor/distributed_gpu_executor.py:149]: @rkooo567 I will check but I'm not sure that this is necessary. At least in the async case there's already an overall per-step timeout that would cover this I think.

**Comment by njhill** [comment]
@rkooo567 OK the spec decoding thing was a tiny fix. And I have added some more detail to the code comments per your latest review.

Thanks @cadedaniel for reviewing too, I was going to ping you to ask for review of that part once the tests were passing :)

@andoorve I can help to look at how we can adapt it to work with PP but it sounds like there is some urgency to get this PR merged first (for the upcoming release?)

**Comment by andoorve** [comment]
Hi @njhill,

Got it, don't want to delay this PR, but just wanted to see if there was any way to loosen this assumption (some mechanism or fallback to per-step task scheduling) that only the rank 0 worker is sending over RPC and receiving results.

The main idea of having other workers (non-drivers) in a task loop is thankfully compatible with PP, but it's just this assumption that there's only 1 parallel group that could cause quite a bit of friction. Happy to chat offline as well.

**Comment by rkooo567** [comment]
@njhill thank you so much for the quick fix!

**Comment by njhill** [comment]
@andoorve I will look more closely at the PP PR tomorrow (and we could chat too if you're free then). Based on what you said and a high level understanding/assumption about how that works I'm fairly confident we can solve it without too much effort.

**Comment by Infinite666** [comment]
Hi @njhill, I am trying to build a new feature that requires calling remote workers to execute methods during this ongoing execution loop. But other remote calls will hang unless the loop stops. Do you have any clean suggestions to solve this? thx!

**Comment by njhill** [comment]
@Infinite666 yes this is a known issue - other calls are held up until there's a gap in request processing. However, it only applies to the V0 code path which we are moving away from as fast as possible. Are you using the latest version of vLLM and can you use V1?

If not, there's some discussion about this and potential fixes for it [here](https://github.com/vllm-project/vllm/issues/12032#issuecomment-2755985601).

**Comment by Infinite666** [comment]
@njhill Thanks for replying! I am currently using v0 code of v0.7, I will take a look at the potential fix and v1 code.
